src/formatDate.js

```diff
@@ -50,7 +50,57 @@
  */
 
 function formatDate(date, fromFormat, toFormat) {
-  // write code here
+  const dateMap = initDateMap(date, fromFormat);
+  let formatedDate = String();
+  const toSeparator = toFormat[3];
+
+  for (let i = 0; i < toFormat.length - 1; i++) {
+    formatedDate += getDatePart(toFormat[i], dateMap) + toSeparator;
+  }
+
+  return formatedDate.substr(0, formatedDate.length - 1);
+}
+
+function initDateMap(date, fromFormat) {
+  const rawDate = getNoSeparatorDate(date, fromFormat);
+  const dateMap = new Map();
+  let currentLength = 0;
+
+  for (let i = 0; i < fromFormat.length - 1; i++) {
+    const leftIndex = currentLength;
+    const rightIndex = leftIndex + fromFormat[i].length;
+    const typeKey = fromFormat[i].charAt(0);
+
+    dateMap.set(typeKey, rawDate.substring(leftIndex, rightIndex));
+    currentLength += fromFormat[i].length;
+  }
+
+  return dateMap;
+}
+
+function getDatePart(dateMapping, map) {
+  const key = dateMapping.charAt(0);
+
+  if (key === 'Y') {
+    return convertYear(map.get(key), dateMapping);
+  }
+
+  return map.get(key);
+}
+
```

Comment:
Cложна
- сначала можно циклом пройтись по fromFormat и создать объект в котором будет день, месяц и год

```
date = {
  day: '10',
  month: '2',
  year: '2020',
}
```
После этого циклом пройтись по toFormat и искать сответствующее поле в объекте date
```
for(let i = 0; i < toFormat.length; i++) {
  switch(toFormat[i]) {
    case 'YY':
    case 'YYYY':
      // добавляем в результат нужную часть из объекта date
      break;
    case 'DD':
      // и так все три части
```

Reply:
changed according to suggestion, with left / right indexes it was overcomplicated from my side, split looks much better. But as far as I know switch case is considered to be a bad practice because it usually violates OCP from SOLID. In this particular case, if I want to widen date format scenarios - I need to refactor switch case construction manually each time so my code is opened for modification. That's why map still looks more flexible to me.